EIPS/eip-5656.md

```diff
@@ -0,0 +1,104 @@
+it can be efficiently stored using `MSTORE8`.
+
+As example use case, copying 256 bytes costs:
+- at least 757 gas pre-EIP-2929 using the identity precompile
+- at least 157 gas post-EIP-2929 using the identity precompile
+- at least 96 gas using unrolled `MLOAD`/`MSTORE` instructions
+- 27 gas using this EIP
+
+According to an analysis of blocks 10537502 to 10538702, roughly 10.5% of memory copies would have had improved performance with the
+availability of an `MCOPY` instruction.
+
+Memory copying is used by languages like Solidity and Vyper, where we expect this improvement to provide efficient means of building
+data structures, including efficient sliced access and copies of memory objects. Having a dedicated `MCOPY` instruction would also add
+forward protection against future gas cost changes to `CALL` instructions in general.
+
+Having a special `MCOPY` instruction makes the job of static analyzers and optimizers easier, since the effects of a `CALL` in general
+have to be fenced, whereas an `MCOPY` instruction would be known to only have memory effects. Even if special cases are added
+for precompiles, a future hard fork could change `CALL` effects, and so any analysis of code using the identity precompile would only
+be valid for a certain range of blocks.
+
+Finally, we expect memory copying to be immensely useful for various computationally heavy operations, such as EVM384,
+where it is identified as a significant overhead.
+
+## Specification
+
+The instruction `MCOPY` is introduced at `0x5c`.
+
+### Input stack
+
+| Stack | Value |
+|-------|-------|
+| top - 0 | `dst` |
+| top - 1 | `src` |
+| top - 2 | `length` |
+
+This ordering matches the other copying instructions, i.e. `CALLDATACOPY`, `RETURNDATACOPY`.
+
+### Output stack
+
+This instructions returns no stack items.
+
+### Semantics
+
+It copies `length` bytes from the offset pointed at `src` to the offset pointed at `dst` in memory.
+Copying takes place as if an intermediate buffer were used, allowing the destination and source to overlap.
+
+If `length > 0` and (`src + length` or `dst + length`) is beyond the current memory length, the memory is extended with respective gas cost applied.
+
+The gas cost of this instruction mirrors that of other `Wcopy` instructions and is `Gverylow + Gcopy * ceil(length / 32)`.
+
+## Rationale
+
+Production implementation of exact-word memory copying and partial-word memory copying can be found in the Solidity, Vyper and Fe compilers.
+
+With [EIP-2929](./eip-2929.md) the call overhead using the identity precompile was reduced from 700 to 100 gas.
+This is still prohibitive for making the precompile a reasonable alternative again.
+
+## Backwards Compatibility
+
+This EIP introduces a new instruction which did not exists previously. Already deployed contracts using this instruction could change their behaviour after this EIP.
+
+## Test Cases
+
+TBA
+
+## Security Considerations
+
+TBA
```

Comment:
```suggestion
Needs discussion.
```
My personal preference. Feel free to ignore it.

Reply:
I think EIP-1 demands a security discussion, if only to explain why there are no impacts.

Reply:
This is correct, but the restrictions are more relaxed for drafts.